Remove time when removing stops

mattwigway opened the pull request:
This also adds the ability to return warnings to the frontend.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/Scenario.java:
```diff
@@ -113,7 +117,16 @@ public TransportNetwork applyToTransportNetwork (TransportNetwork originalNetwor
         }
+
+        copiedNetwork.scenarioApplicationWarnings = modificationsWithWarnings.stream()
+                .map(m -> new TaskError(m, m.warnings))
```
It seems like this constructor should take one object of type Modification (since the second argument is implicit in the first), and then we should call  `modificationsWithWarnings.stream().map(TaskError::new)`

mattwigway replied:
The second argument is actually not implicit in the first, because a task error can describe either the errors or the warnings.

abyrd commented on src/main/java/com/conveyal/r5/analyst/broker/BrokerHttpHandler.java:
```diff
@@ -65,6 +66,10 @@ public void service(Request request, Response response) throws Exception {
         if (pathComponents.length < 2) {
             response.setStatus(HttpStatus.BAD_REQUEST_400);
             response.setDetailMessage("path should have at least one part");
+            TaskError error = new TaskError(null, "Path should have at least one part", "");
```
This would usually be caused by an error in how our components communicate with each other, so I'm not sure we need to wrap the error message in a data model object that's intended to show user-oriented errors.

abyrd replied:
Then again I see that you've done this throughout the class, so I guess the idea is to change the purpose/meaning of TaskError to represent all errors.

mattwigway replied:
The reason I did this is because anything that causes a 400 error is expected by the frontend to have a JSON-formatted error payload so we can display the error.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/Modification.java:
```diff
@@ -65,7 +64,14 @@ public final void setType (String type) {
      * Modification should not be applied, and this Set should contain Strings describing all the problems.
      */
-    public final Set<String> warnings = new HashSet<String>();
+    public transient final Set<String> errors = new HashSet<>();
```
I'm glad to see the name of this field change, it was inaccurate.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/RemoveStops.java:
```diff
@@ -47,6 +48,16 @@
 
+    /**
+     * Number of seconds (in addition to dwell time, if any) to remove from the schedule at each removed stop,
+     * to account for acceleration, deceleration, merging, and so on.
```
Maybe we should say "vehicle acceleration and deceleration" and "merging into traffic" to make it clearer that we're talking about time saved by physical changes in bus operations and not something in code.

mattwigway replied:
Sure, although I'm not sure what code operation is implied by acceleration :)

mattwigway replied:
And of course merging into traffic could be confusing once we have a branch called traffic :) :)

mattwigway replied:
But seriously, thanks, I updated the comment.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/RemoveStops.java:
```diff
@@ -47,6 +48,16 @@
+     * Number of seconds (in addition to dwell time, if any) to remove from the schedule at each removed stop,
+     * to account for acceleration, deceleration, merging, and so on.
+     *
+     * If removing this time would make the hop time between two stops that remain <= 0, an error will be surfaced.
```
It's actually a warning that's surfaced in the UI.

mattwigway replied:
Fixed.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/RemoveStops.java:
```diff
@@ -147,23 +158,63 @@ public TripPattern processTripPattern (TripPattern originalTripPattern) {
+                    accumulatedRideTime += rideTime;
+                    int secondsToRemove = secondsSavedAtEachStop * nStopsRemovedSinceLastStop;
+                    if (nStopsRemovedSinceLastStop > 0 && accumulatedRideTime < secondsToRemove) {
+                        // figure out which stops are causing the problem
```
The comment says we're finding stops that cause a problem, but doesn't say that this conditional block handles a problem, or what that problem is.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/RemoveStops.java:
```diff
@@ -147,23 +158,63 @@ public TripPattern processTripPattern (TripPattern originalTripPattern) {
+                    // all the time savings before the stop. After we've accumulated ride time for a whole segment,
+                    // we check to make sure it's positive.
                     accumulatedRideTime += rideTime;
+                    nStopsRemovedSinceLastStop++;
```
It seems odd to me to count how many stops have been removed and then multiply that by the number of stops removed. Shouldn't it be possible to subtract secondsSavedAtEachStop out of accumulatedRideTime at each iteration, and just keep an eye on the accumulated total?

mattwigway replied:
Yes, we could do it that way. I didn't because we need to know the number of stops removed so we can generate the warning messages, but there wouldn't be any harm in just keeping that separate from the actual calculation.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/RemoveStops.java:
```diff
@@ -147,23 +158,63 @@ public TripPattern processTripPattern (TripPattern originalTripPattern) {
+                        }
+
+                        // map them back to names
+                        String problemStopNames = IntStream.of(problemStops)
```
Having the error handling code inserted in the conditional body inside the loop stretches out the blocks and makes it a bit harder to see the overall flow. It seems sufficient to just record all the offending stops in a set and handle the error generation outside the inner blocks, potentially even in a function.

abyrd commented on src/main/java/com/conveyal/r5/analyst/scenario/Scenario.java:
```diff
@@ -93,11 +95,13 @@ public TransportNetwork applyToTransportNetwork (TransportNetwork originalNetwor
         // Check all the parameters before applying any modifications.
         // Might some parameters may become valid/invalid because of previous modifications in the list?
         List<Modification> badModifications = new ArrayList<>();
```
Parallelism: if the next field is called modificationsWithWarnings maybe this one should be renamed modificationsWithErrors.

mattwigway replied:
Fixed.

mattwigway commented:
@abyrd I believe I have addressed your concerns, please re-review. I did not change how the time removed get subtracted from accumulated ride time; see my response to your comment and let me know if you'd still like me to make that change.

abyrd commented:
@mattwigway thanks, at the end of the day I suppose it's a question of style and the functionality is needed soon, so let's just merge it as is.
